File: 39- Longest common substring/Java/solution3.java
// By using dynamic programming:

// Time complexity: O(nm)
// Space complexity: O(nm)

static int lcs(String str1, String str2){
  int n = str1.length();
  int m = str2.length();
  int[][] dp = new int[n+1][m+1];
  int maxLength = 0;
  for(int i = 1; i < n+1; i++){
    for(int j = 1; j < m+1; j++){
      if(str1.charAt(i-1) == str2.charAt(j-1)){
        dp[i][j] = 1 + dp[i-1][j-1];
        maxLength = Math.max(maxLength, dp[i][j]);
      }else{
        dp[i][j] = 0;
      } 
    }
  }
  return maxLength;
}

File: 39- Longest common substring/Python/solution3.py
# By using dynamic programming:

# Time complexity: O(nm)
# Space complexity: O(nm)


def lcs(str1, str2):
  n = len(str1)
  m = len(str2)
  dp = [[0] * (m+1) for i in range(n+1)]
  maxLength = 0
  for i in range(1, n+1):
    for j in range(1, m+1):
      if str1[i-1] == str2[j-1]:
        dp[i][j] = 1 + dp[i-1][j-1]
        maxLength = max(maxLength, dp[i][j])
      else:
        dp[i][j] = 0
  return maxLength


